File: apps/android/app/src/main/java/dating/swipe/core/DatingCoreBridge.kt
package dating.swipe.core

/**
 * Facade over audited UniFFI exports.
 * Uses generated bindings when JNI lib is present; otherwise STAGING_FALLBACK stub.
 */
object DatingCoreBridge {
    const val STAGING_FALLBACK_LABEL = "STAGING_FALLBACK"

    val isStagingFallback: Boolean
        get() = !UniffiCore.isNativeLibraryLoaded()

    fun protocolVersion(): UShort = UniffiCore.protocolVersion()

    data class EligibilitySummary(
        val adult: Boolean,
        val ageBand: String?,
        val issuedAtUnix: Long,
        val expiresAtUnix: Long,
        val provider: String,
        val appealAllowed: Boolean,
    )

    sealed class EligibilityError : Exception() {
        data object Ineligible : EligibilityError()
        data object Ambiguous : EligibilityError()
        data object Expired : EligibilityError()
        data object Revoked : EligibilityError()
        data object ProviderUnavailable : EligibilityError()
    }

    @Throws(EligibilityError::class)
    fun evaluateMockAgeEligibility(
        adult: Boolean,
        ambiguous: Boolean,
        unavailable: Boolean,
    ): EligibilitySummary = UniffiCore.evaluateMockAgeEligibility(adult, ambiguous, unavailable)

    @Throws(EligibilityError::class)
    fun assertDiscoveryAllowed(summary: EligibilitySummary, nowUnix: Long) {
        UniffiCore.assertDiscoveryAllowed(summary, nowUnix)
    }
}

/**
 * Dispatches to generated UniFFI Kotlin or STAGING_FALLBACK stubs matching UniFFI names.
 */
private object UniffiCore {
    fun isNativeLibraryLoaded(): Boolean = runCatching {
        uniffi.dating_uniffi_bindings.uniffiEnsureInitialized()
        true
    }.getOrDefault(false)

    fun protocolVersion(): UShort {
        if (isNativeLibraryLoaded()) {
            return uniffi.dating_uniffi_bindings.`protocolVersion`()
        }
        return 1u
    }

    @Throws(DatingCoreBridge.EligibilityError::class)
    fun evaluateMockAgeEligibility(
        adult: Boolean,
        ambiguous: Boolean,
        unavailable: Boolean,
    ): DatingCoreBridge.EligibilitySummary {
        if (isNativeLibraryLoaded()) {
            return try {
                val summary = uniffi.dating_uniffi_bindings.`evaluateMockAgeEligibility`(
                    adult,
                    ambiguous,
                    unavailable,
                )
                DatingCoreBridge.EligibilitySummary(
                    adult = summary.adult,
                    ageBand = summary.ageBand,
                    issuedAtUnix = summary.issuedAtUnix,
                    expiresAtUnix = summary.expiresAtUnix,
                    provider = summary.provider,
                    appealAllowed = summary.appealAllowed,
                )
            } catch (e: uniffi.dating_uniffi_bindings.EligibilityErrorCode) {
                throw mapEligibilityError(e)
            }
        }

        return UniffiStagingFallback.evaluateMockAgeEligibility(adult, ambiguous, unavailable)
    }

    @Throws(DatingCoreBridge.EligibilityError::class)
    fun assertDiscoveryAllowed(summary: DatingCoreBridge.EligibilitySummary, nowUnix: Long) {
        if (isNativeLibraryLoaded()) {
            try {
                val ffiSummary = uniffi.dating_uniffi_bindings.EligibilitySummary(
                    adult = summary.adult,
                    ageBand = summary.ageBand,
                    issuedAtUnix = summary.issuedAtUnix,
                    expiresAtUnix = summary.expiresAtUnix,
                    provider = summary.provider,
                    appealAllowed = summary.appealAllowed,
                )
                uniffi.dating_uniffi_bindings.`assertDiscoveryAllowed`(ffiSummary, nowUnix)
            } catch (e: uniffi.dating_uniffi_bindings.EligibilityErrorCode) {
                throw mapEligibilityError(e)
            }
            return
        }
        UniffiStagingFallback.assertDiscoveryAllowed(summary, nowUnix)
    }

    private fun mapEligibilityError(
        e: uniffi.dating_uniffi_bindings.EligibilityErrorCode,
    ): DatingCoreBridge.EligibilityError = when (e) {
        is uniffi.dating_uniffi_bindings.EligibilityErrorCode.Ineligible ->
            DatingCoreBridge.EligibilityError.Ineligible
        is uniffi.dating_uniffi_bindings.EligibilityErrorCode.Ambiguous ->
            DatingCoreBridge.EligibilityError.Ambiguous
        is uniffi.dating_uniffi_bindings.EligibilityErrorCode.Expired ->
            DatingCoreBridge.EligibilityError.Expired
        is uniffi.dating_uniffi_bindings.EligibilityErrorCode.Revoked ->
            DatingCoreBridge.EligibilityError.Revoked
        is uniffi.dating_uniffi_bindings.EligibilityErrorCode.ProviderUnavailable ->
            DatingCoreBridge.EligibilityError.ProviderUnavailable
    }
}

/**
 * Pure-Kotlin STAGING_FALLBACK matching UniFFI export names (no JNI).
 */
internal object UniffiStagingFallback {
    fun protocolVersion(): UShort = 1u

    @Throws(DatingCoreBridge.EligibilityError::class)
    fun evaluateMockAgeEligibility(
        adult: Boolean,
        ambiguous: Boolean,
        unavailable: Boolean,
    ): DatingCoreBridge.EligibilitySummary {
        if (unavailable) throw DatingCoreBridge.EligibilityError.ProviderUnavailable
        if (ambiguous) throw DatingCoreBridge.EligibilityError.Ambiguous
        if (!adult) throw DatingCoreBridge.EligibilityError.Ineligible
        val now = System.currentTimeMillis() / 1000
        return DatingCoreBridge.EligibilitySummary(
            adult = true,
            ageBand = "18+",
            issuedAtUnix = now,
            expiresAtUnix = now + 86_400,
            provider = "mock-${DatingCoreBridge.STAGING_FALLBACK_LABEL}",
            appealAllowed = true,
        )
    }

    @Throws(DatingCoreBridge.EligibilityError::class)
    fun assertDiscoveryAllowed(summary: DatingCoreBridge.EligibilitySummary, nowUnix: Long) {
        if (!summary.adult) throw DatingCoreBridge.EligibilityError.Ineligible
        if (nowUnix >= summary.expiresAtUnix) throw DatingCoreBridge.EligibilityError.Expired
    }
}

File: apps/ios/Sources/DatingCoreBridge.swift
import Foundation

/// Mobile facade over audited UniFFI exports.
/// Xcode app builds define `DATING_UNIFFI_LINKED` and compile `Generated/`.
/// SPM-only builds without the native lib keep the STAGING mock below.
public enum DatingCoreBridge {
    public static var isStagingFallback: Bool {
        #if DATING_UNIFFI_LINKED
        return false
        #else
        return true
        #endif
    }

    public static func protocolVersion() -> UInt16 {
        #if DATING_UNIFFI_LINKED
        return uniffi_protocol_version()
        #else
        return 1
        #endif
    }

    public struct EligibilitySummary: Equatable, Sendable {
        public let adult: Bool
        public let ageBand: String?
        public let issuedAtUnix: Int64
        public let expiresAtUnix: Int64
        public let provider: String
        public let appealAllowed: Bool

        public init(
            adult: Bool,
            ageBand: String?,
            issuedAtUnix: Int64,
            expiresAtUnix: Int64,
            provider: String,
            appealAllowed: Bool
        ) {
            self.adult = adult
            self.ageBand = ageBand
            self.issuedAtUnix = issuedAtUnix
            self.expiresAtUnix = expiresAtUnix
            self.provider = provider
            self.appealAllowed = appealAllowed
        }
    }

    public enum EligibilityError: Error, Equatable {
        case ineligible
        case ambiguous
        case expired
        case revoked
        case providerUnavailable
    }

    public static func evaluateMockAgeEligibility(
        adult: Bool,
        ambiguous: Bool,
        unavailable: Bool
    ) throws -> EligibilitySummary {
        #if DATING_UNIFFI_LINKED
        do {
            let summary = try uniffi_evaluate_mock_age_eligibility(
                adult: adult,
                ambiguous: ambiguous,
                unavailable: unavailable
            )
            return EligibilitySummary(
                adult: summary.adult,
                ageBand: summary.ageBand,
                issuedAtUnix: summary.issuedAtUnix,
                expiresAtUnix: summary.expiresAtUnix,
                provider: summary.provider,
                appealAllowed: summary.appealAllowed
            )
        } catch let err as EligibilityErrorCode {
            throw mapEligibility(err)
        }
        #else
        if unavailable {
            throw EligibilityError.providerUnavailable
        }
        if ambiguous {
            throw EligibilityError.ambiguous
        }
        if !adult {
            throw EligibilityError.ineligible
        }
        let now = Int64(Date().timeIntervalSince1970)
        return EligibilitySummary(
            adult: true,
            ageBand: "18+",
            issuedAtUnix: now,
            expiresAtUnix: now + 86_400,
            provider: "mock-STAGING",
            appealAllowed: true
        )
        #endif
    }

    public static func assertDiscoveryAllowed(
        _ summary: EligibilitySummary,
        nowUnix: Int64
    ) throws {
        #if DATING_UNIFFI_LINKED
        do {
            try uniffi_assert_discovery_allowed_bridge(
                adult: summary.adult,
                ageBand: summary.ageBand,
                issuedAtUnix: summary.issuedAtUnix,
                expiresAtUnix: summary.expiresAtUnix,
                provider: summary.provider,
                appealAllowed: summary.appealAllowed,
                nowUnix: nowUnix
            )
        } catch let err as EligibilityErrorCode {
            throw mapEligibility(err)
        }
        #else
        guard summary.adult else {
            throw EligibilityError.ineligible
        }
        guard nowUnix < summary.expiresAtUnix else {
            throw EligibilityError.expired
        }
        #endif
    }

    public static func coarseRegionBandLabel(lat: Double, lon: Double, jitterSeed: UInt64) -> String {
        #if DATING_UNIFFI_LINKED
        return uniffi_coarse_region_band_label(lat: lat, lon: lon, jitterSeed: jitterSeed)
        #else
        _ = (lat, lon, jitterSeed)
        return "nearby"
        #endif
    }

    public static func stagingProfileIdBytes(label: String) -> Data {
        #if DATING_UNIFFI_LINKED
        return uniffi_staging_profile_id_bytes(label: label)
        #else
        var bytes = Array(label.utf8)
        if bytes.count < 32 {
            bytes.append(contentsOf: repeatElement(UInt8(0), count: 32 - bytes.count))
        }
        return Data(bytes.prefix(32))
        #endif
    }

    public static func generateLocalIdentity() -> (profileIdHex: String, rootPublicKeyHex: String)? {
        #if DATING_UNIFFI_LINKED
        // Prefer retainHandle() for presence signing; this summary helper remains for SPM/mock.
        let handle = uniffi_generate_local_identity()
        let summary = handle.publicIdentitySummary()
        return (summary.profileIdHex, summary.rootPublicKeyHex)
        #else
        return nil
        #endif
    }

    #if DATING_UNIFFI_LINKED
    public static func retainIdentityHandle() -> IdentityHandle {
        uniffi_generate_local_identity()
    }
    #endif

    /// Local match/dislike engine backed by audited UniFFI store when linked.
    public final class LocalMatchEngine {
        #if DATING_UNIFFI_LINKED
        private let store: AuditedMatchStore = uniffi_make_match_store()
        #endif

        public init() {}

        public func pass(profileLabel: String) throws {
            let id = DatingCoreBridge.stagingProfileIdBytes(label: profileLabel)
            #if DATING_UNIFFI_LINKED
            try store.recordDislike(profileId: id)
            #else
            _ = id
            #endif
        }

        public func like(profileLabel: String) throws {
            let id = DatingCoreBridge.stagingProfileIdBytes(label: profileLabel)
            #if DATING_UNIFFI_LINKED
            try store.recordLike(profileId: id)
            #else
            _ = id
            #endif
        }

        public func confirmStagingMatch(profileLabel: String) throws {
            let id = DatingCoreBridge.stagingProfileIdBytes(label: profileLabel)
            #if DATING_UNIFFI_LINKED
            try store.confirmStagingMatch(profileId: id)
            #else
            _ = id
            #endif
        }

        public func block(profileLabel: String) throws {
            let id = DatingCoreBridge.stagingProfileIdBytes(label: profileLabel)
            #if DATING_UNIFFI_LINKED
            try store.blockStaging(profileId: id)
            #else
            _ = id
            #endif
        }

        public func stateLabel(profileLabel: String) -> String {
            let id = DatingCoreBridge.stagingProfileIdBytes(label: profileLabel)
            #if DATING_UNIFFI_LINKED
            return (try? store.matchStateLabel(profileId: id)) ?? "unknown"
            #else
            _ = id
            return "synthetic"
            #endif
        }
    }

    #if DATING_UNIFFI_LINKED
    private static func mapEligibility(_ err: EligibilityErrorCode) -> EligibilityError {
        switch err {
        case .Ineligible: return .ineligible
        case .Ambiguous: return .ambiguous
        case .Expired: return .expired
        case .Revoked: return .revoked
        case .ProviderUnavailable: return .providerUnavailable
        }
    }
    #endif
}
